### PR TITLE
dependencies messages, no curl on win32, fixes to min Version

### PR DESCRIPTION
**Pull request by ranasaria**
This PR changes the following:
1. Remove curl dependency on win32.
2. Informatonal messages for dependency on tools that we assume are pre-installed.
3. Change "..." to … (ellipsis character)
4. Change to ensure that minimum version is met post tools installation.
5. Column resizes in tools table.
6. In toolsService the tools array changed to a Map to enable quick lookup by tool name.

### CODE REVIEW

File: extensions/resource-deployment/src/services/tools/azdataTool.ts [hunk 1]
```diff
@@ -112,38 +112,47 @@ export class AzdataTool extends ToolBase {
 	private get azdataInstallAdditionalArgs(): string {
 		return vscode.workspace.getConfiguration(DeploymentConfigurationKey)[azdataPipInstallArgsKey];
 	}
+
+	// Additional Informational messages for various OsTypes.
+	protected additionalInformation: Map<OsType, InformationalMessageType[]> = new Map<OsType, InformationalMessageType[]>([
+		[OsType.linux, []],
```
Comment (alanrenmsft) [inline extensions/resource-deployment/src/services/tools/azdataTool.ts]: as of now, linux also depends on python and pip3

File: extensions/resource-deployment/src/services/tools/toolBase.ts [hunk 2]
```diff
@@ -14,7 +14,7 @@ import { IPlatformService } from '../platformService';
 const localize = nls.loadMessageBundle();
 const toolStatusNotInstalled: string = localize('deploymentDialog.ToolStatus.NotInstalled', "Not Installed");
 const toolStatusInstalled: string = localize('deploymentDialog.ToolStatus.Installed', "Installed");
-const toolStatusInstalling: string = localize('deploymentDialog.ToolStatus.Installing', "Installing");
+const toolStatusInstalling: string = localize('deploymentDialog.ToolStatus.Installing', "Installing …");
```
Comment (alanrenmsft) [inline extensions/resource-deployment/src/services/tools/toolBase.ts]: should there be a space in between?

File: extensions/resource-deployment/src/services/tools/toolBase.ts [hunk 3]
```diff
@@ -26,6 +26,22 @@ const toolStatusLocalized: Map<ToolStatus, string> = new Map<ToolStatus, string>
 	[ToolStatus.Failed, toolStatusFailed]
 ]);
 
+export const enum InformationalMessageType {
```
Comment (alanrenmsft) [inline extensions/resource-deployment/src/services/tools/toolBase.ts]: I'd prefer to make it specific indicating dependencies instead of general informationalMessageType

File: extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts [hunk 4]
```diff
@@ -196,35 +195,25 @@ export class ResourceTypePickerDialog extends DialogBase {
 	}
 
 	private updateToolsDisplayTable(): void {
-		this.toolRefreshTimestamp = new Date().getTime();
-		const currentRefreshTimestamp = this.toolRefreshTimestamp;
```
Comment (alanrenmsft) [inline extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts]: I added this to avoid the information being displayed when the selection has changed, please don't remove the related code. and looks like you are trying to improve the response time of the table?

Comment (ranasaria) [reply extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts]: fixed

File: extensions/resource-deployment/src/services/tools/toolBase.ts (hunk 2 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/toolBase.ts]: removed

File: extensions/resource-deployment/src/services/tools/toolBase.ts (hunk 3 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/toolBase.ts]: changed.

File: extensions/resource-deployment/src/services/tools/azdataTool.ts (hunk 1 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/azdataTool.ts]: fixed

File: extensions/resource-deployment/src/services/tools/azdataTool.ts [hunk 5]
```diff
@@ -112,38 +112,46 @@ export class AzdataTool extends ToolBase {
...
 		command: 'apt-get install gnupg ca-certificates curl apt-transport-https lsb-release -y'
 	},
 	{
 		sudo: true,
-		comment: localize('resourceDeployment.Azdata.DownloadAndInstallingSigningKey', "downloading and installing the signing key for azdata ..."),
+		comment: localize('resourceDeployment.Azdata.DownloadAndInstallingSigningKey', "downloading and installing the signing key for azdata …"),
 		command: 'wget -qO- https://packages.microsoft.com/keys/microsoft.asc | apt-key add -'
 	},
+	// Double check with Sean Wells on the current url for azdata corresponding to https://packages.microsoft.com/config/ubuntu/16.04/mssql-server-preview.list)
 	{
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/services/tools/azdataTool.ts]: Why is this here? Tasks should be tracked somewhere else, not in code

File: extensions/resource-deployment/src/services/tools/kubeCtlTool.ts [hunk 6]
```diff
@@ -75,86 +75,93 @@ export class KubeCtlTool extends ToolBase {
...
-		comment: localize('resourceDeployment.Kubectl.DeletePreviousDownloadedKubectl.exe', "deleting previously downloaded kubectl.exe if one exists ..."),
-		command: `IF EXIST .\kubectl.exe DEL /F .\kubectl.exe`,
+		comment: localize('resourceDeployment.Kubectl.DeletePreviousDownloadedKubectl.exe', "deleting previously downloaded kubectl.exe if one exists …"),
+		command: 'IF EXIST .\\kubectl.exe DEL /F .\\kubectl.exe',
 	},
 	{
-		comment: localize('resourceDeployment.Kubectl.DownloadingAndInstallingKubectl', "downloading and installing the latest kubectl.exe ..."),
-		command: `for /f %i in ('curl -s https://storage.googleapis.com/kubernetes-release/release/stable.txt') do curl -LO https://storage.googleapis.com/kubernetes-release/release/%i/bin/windows/amd64/kubectl.exe`
+		comment: localize('resourceDeployment.Kubectl.DownloadingAndInstallingKubectl', "downloading and installing the latest kubectl.exe …"),
+		command: `powershell -Command "& {(New-Object System.Net.WebClient).DownloadFile('https://storage.googleapis.com/kubernetes-release/release/v1.16.2/bin/windows/amd64/kubectl.exe', 'kubectl.exe')}"`
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/services/tools/kubeCtlTool.ts]: Hm, this isn't really going to necessarily be true. If we do truly want to always get the latest then the suggestion on their page is to read [https://storage.googleapis.com/kubernetes-release/release/stable.txt](https://storage.googleapis.com/kubernetes-release/release/stable.txt) to find the latest stable version

Comment (Charles-Gagnon) [reply extensions/resource-deployment/src/services/tools/kubeCtlTool.ts]: Oh yeah you already did this with the old curl version. So I'd fix this to do the same thing

---
In reply to: [339315886](https://github.com/microsoft/azuredatastudio/pull/8039#discussion_r339315886) [](ancestors = 339315886)

File: extensions/resource-deployment/src/services/tools/toolBase.ts [hunk 7]
```diff
@@ -26,6 +26,22 @@ const toolStatusLocalized: Map<ToolStatus, string> = new Map<ToolStatus, string>
 	[ToolStatus.Failed, toolStatusFailed]
 ]);
 
+export const enum dependencyType {
+	PythonAndPip3 = 'PythonAndPip3',
+	Brew = 'Brew',
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/services/tools/toolBase.ts]: My preference would be to split these into separate dependencies and then or them together. I could forsee some tool needing python as a dependency but not necessarily pip

File: extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts [hunk 8]
```diff
@@ -323,16 +334,26 @@ export class ResourceTypePickerDialog extends DialogBase {
 
 	private async installTools(): Promise<void> {
 		this._installToolButton.enabled = false;
-		let i: number = 0;
+		let tool: ITool;
 		try {
-			for (; i < this._tools.length; i++) {
-				if (this._tools[i].needsInstallation) {
+			const toolRequirements = this.getCurrentProvider().requiredTools;
+			for (let i: number = 0; i < toolRequirements.length; i++) {
```
Comment (Charles-Gagnon) [inline extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts]: You changed the logic here now so that the index is from the toolRequirements array but didn't update all the dialog messages below (such as line 347). Make sure you go through and fix those to use the appropriate tool reference.

File: extensions/resource-deployment/src/services/tools/kubeCtlTool.ts (hunk 6 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/kubeCtlTool.ts]: No - Fixing the version was a conscious independent choice by Alan as we wanted it to be predictable.

File: extensions/resource-deployment/src/services/tools/toolBase.ts (hunk 7 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/toolBase.ts]: Sure we will do that when we get to that tool. Today they are needed together in tandem. In fact in future our direction is to move away from them with dedicated installers that do not have any dependencies.

File: extensions/resource-deployment/src/services/tools/azdataTool.ts (hunk 5 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/azdataTool.ts]: will remove the comment

File: extensions/resource-deployment/src/services/tools/kubeCtlTool.ts (hunk 6 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/kubeCtlTool.ts]: I will change code install version v1.16.2 across all platforms.

File: extensions/resource-deployment/src/services/tools/kubeCtlTool.ts [hunk 9]
```diff
@@ -75,86 +75,93 @@ export class KubeCtlTool extends ToolBase {
...
 		command: 'apt-get update'
 	},
 	{
 		sudo: true,
-		comment: localize('resourceDeployment.Kubectl.InstallingKubectl', "installing kubectl ..."),
+		comment: localize('resourceDeployment.Kubectl.InstallingKubectl', "installing kubectl …"),
 		command: 'apt-get install -y kubectl'
 	}
 ];
 // TODO: Remove dependency on curl on Win32 and use powershell Invoke-WebRequest instead
```
Comment (alanrenmsft) [inline extensions/resource-deployment/src/services/tools/kubeCtlTool.ts]: now you can remove this comment, right?

File: extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts [hunk 10]
```diff
@@ -323,16 +334,26 @@ export class ResourceTypePickerDialog extends DialogBase {
...
+				const toolReq = toolRequirements[i];
+				tool = this.toolsService.getToolByName(toolReq.name)!;
+				if (tool.needsInstallation) {
 					// Update the informational message
 					this._dialogObject.message = {
 						level: azdata.window.MessageLevel.Information,
 						text: localize('deploymentDialog.InstallingTool', "Required tool '{0}' [ {1} ] is being installed now.", this._tools[i].displayName, this._tools[i].homePage)
 					};
 					await this._tools[i].install();
+					if (tool.isInstalled && toolReq.version && !tool.isSameOrNewerThan(toolReq.version)) {
```
Comment (alanrenmsft) [inline extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts]: hmm, this shouldn't happen, should ony be a dev time scenario, you updated the required version but not the install tool method, you can treat it as a dev time error.

File: extensions/resource-deployment/src/services/tools/azdataTool.ts (hunk 5 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/azdataTool.ts]: done

File: extensions/resource-deployment/src/services/tools/kubeCtlTool.ts (hunk 6 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/kubeCtlTool.ts]: no further investigation since brew install on mac and apt-get install on linux are pulling latest version, so I am changing to pull the latest version of kubectl for windows as well.

File: extensions/resource-deployment/src/services/tools/kubeCtlTool.ts (hunk 9 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/kubeCtlTool.ts]: yes will do.

File: extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts (hunk 10 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts]: yeah perhaps the check ensures that we still do have the correct version before proceeding so it is a good defensive code.

File: extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts (hunk 8 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/ui/resourceTypePickerDialog.ts]: Good catch. Will do the necessary update

File: extensions/resource-deployment/src/services/tools/kubeCtlTool.ts (hunk 6 above)
Comment (ranasaria) [reply extensions/resource-deployment/src/services/tools/kubeCtlTool.ts]: Fixed to make new powershell implementation of what the curl based one was before.
